applications/CompressiblePotentialFlowApplication/custom_strategies/transonic_residualbased_incrementalupdate_static_scheme.h

```diff
@@ -0,0 +1,343 @@
+    explicit TransonicResidualBasedIncrementalUpdateStaticScheme()
+        : BaseType()
+    {  
+    }
+
+    /**
+     * @brief Constructor. The pseudo static scheme (parameters)
+     * @param ThisParameters Dummy parameters
+     */
+    explicit TransonicResidualBasedIncrementalUpdateStaticScheme(Parameters ThisParameters) : BaseType()
+    {
+        // Validate and assign defaults
+        ThisParameters = this->ValidateAndAssignParameters(ThisParameters, this->GetDefaultParameters());
+        this->AssignSettings(ThisParameters);
+    }
+
+
+    /** Destructor.
+    */
+    ~TransonicResidualBasedIncrementalUpdateStaticScheme() override {}
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief Create method
+     * @param ThisParameters The configuration parameters
+     */
+    typename BaseSchemeType::Pointer Create(Parameters ThisParameters) const override
+    {
+        return Kratos::make_shared<ClassType>(ThisParameters);
+    }
+
+    /**
+     * @brief This is the place to initialize the Scheme.
+     * @details This is intended to be called just once when the strategy is initialized
+     * @param rModelPart The model part of the problem to solve
+     */
+    void Initialize(ModelPart& rModelPart) override
+    {
+        KRATOS_TRY
+
+        rModelPart.GetProcessInfo()[CRITICAL_MACH]          = mInitialCriticalMach;
+        rModelPart.GetProcessInfo()[UPWIND_FACTOR_CONSTANT] = mInitialUpwindFactorConstant;
+        rModelPart.GetProcessInfo()[MACH_LIMIT]             = std::sqrt(mMachNumberSquaredLimit);
+
+        BaseType::SetSchemeIsInitialized(true);
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief Function to be called when it is needed to initialize an iteration. It is designed to be called at the beginning of each non linear iteration
+     * @note Take care: the elemental function with the same name is NOT called here.
+     * @warning Must be defined in derived classes
+     * @details The function is called in the builder for memory efficiency
+     * @param rModelPart The model part of the problem to solve
+     * @param A LHS matrix
+     * @param Dx Incremental update of primary variables
+     * @param b RHS Vector
+     */
+    void InitializeNonLinIteration(
+        ModelPart& rModelPart,
+        TSystemMatrixType& A,
+        TSystemVectorType& Dx,
+        TSystemVectorType& b
+        ) override
+    {
+        KRATOS_TRY
+
+        // Update the upwind factor constant and critical mach with the user-defined values
+        if (rModelPart.GetProcessInfo()[CONVERGENCE_RATIO] <= mUpdateRelativeResidualNorm &&
+            rModelPart.GetProcessInfo()[NL_ITERATION_NUMBER] > 1                          &&
+            mUpdatedValues == false){
+
+            rModelPart.GetProcessInfo()[CRITICAL_MACH]          = mTargetCriticalMach;
+            rModelPart.GetProcessInfo()[UPWIND_FACTOR_CONSTANT] = mTargetUpwindFactorConstant;
```

Comment:
Aren't we always setting the very same values? In other words, where are `mTargetCriticalMach` and `mTargetUpwindFactorConstant` updated?

Reply:
OK. I think that now I understand. Once a certain tolerance is reached, you enforce the critical mach and upwind factor to a user-defined value, so the solution is less diffusive.

Shouldn't we do this at each non-linear iteration just in case or, in other words, are we sure that these two are computed once and never touched again right?

Reply:
Yes, doing it once is enough since they are not calculated internally.

The _mUpdatedValues_ ​​flag is just to make sure we do it once, although it wouldn't change anything if we did it every step.

Reply:
OK. Up to you but, just in case and taking into account that this is an inexpensive operation, I'd always set them if the residual check is fulfilled (in other words, I'd remove the "do it once" check).